pkg/util/server/error.go

```diff
@@ -60,10 +63,22 @@ func ClientHTTPStatusAndError(err error) (int, error) {
 		return http.StatusBadRequest, err
 	case errors.Is(err, user.ErrNoOrgID):
 		return http.StatusBadRequest, err
+	case isRPC && (int(s.Code())/100 == 4 || int(s.Code())/100 == 5):
+		return int(s.Code()), errors.New(s.Message())
 	default:
 		if grpcErr, ok := httpgrpc.HTTPResponseFromError(err); ok {
 			return int(grpcErr.Code), errors.New(string(grpcErr.Body))
 		}
 		return http.StatusInternalServerError, err
 	}
 }
+
+// WrapError wraps an error in a protobuf status.
+func WrapError(err error) *rpc.Status {
+	code, err := ClientHTTPStatusAndError(err)
+	return status.New(codes.Code(code), err.Error()).Proto()
+}
+
+func UnwrapError(s *rpc.Status) error {
+	return status.ErrorProto(s)
+}
```

Comment:
I think this breaks the retry logic https://github.com/grafana/loki/blob/main/pkg/querier/queryrange/queryrangebase/retry.go#L91-L92 :/

Reply:
We should use an embedded HTTP Response here like in here https://github.com/grafana/loki/blob/3555001aee06f01d26a33880326aa1bb9ab1fc93/pkg/querier/worker/util.go#L114-L117

Reply:
After a talk with @duricanikolic I think we should migrate all errors to the gRPC `Status` interface eventually.

Reply:
TIL: `httpgrpc.Errof` is returning an `httpgrpc.HTTPResponse` inside a `rpc.Status` as an error. So the current solution is fine but all user errors should be changed to `status.Errorf`.